### PR TITLE
feat(plugins): hooks and events can be declared in elgg-plugin.php

### PR DESCRIPTION
**Pull request by jdalsem**

### CODE REVIEW

**Comment by iionly** [comment]
Is there really a reason for removal of the activate.php/deactivate.php functionality? If it gets removed (and I see the suggestion to use the bootstrap class activate/deactivate functions) would it be possible to add some info to the docs?

**Comment by jdalsem** [comment]
> Is there really a reason for removal of the activate.php/deactivate.php functionality

yes, elgg now has to many ways of doing this and we need to reduce the unwanted ways



> If it gets removed (and I see the suggestion to use the bootstrap class activate/deactivate functions) would it be possible to add some info to the docs?

yep, we need some docs for the bootstrap class functions (i created #12691 for that)

File: engine/tests/phpunit/unit/ElggPluginUnitTest.php [hunk 1]
```diff
@@ -93,13 +93,21 @@ public function testCanLoadStaticConfig() {
...
+				'user_default1' => 'set1',
+			],
 		];
 
 		$this->assertEquals($config['entities'], $plugin->getStaticConfig('entities'));
 		$this->assertEquals($config['actions'], $plugin->getStaticConfig('actions'));
 		$this->assertEquals($config['widgets'], $plugin->getStaticConfig('widgets'));
 		$this->assertEquals($config['routes'], $plugin->getStaticConfig('routes'));
+		$this->assertEquals($config['settings'], $plugin->getStaticConfig('settings'));
+		$this->assertEquals($config['user_settings'], $plugin->getStaticConfig('user_settings'));
```
Comment (jeabakker) [inline engine/tests/phpunit/unit/ElggPluginUnitTest.php]: why no checks for `hooks` and `events`

**Review by jeabakker — DISMISSED** [review]
No integration tests to check if (un)registering `hooks` and `events` work?

File: docs/guides/javascript.rst [hunk 2]
```diff
@@ -168,15 +168,17 @@ Setting the URL of a module
 
 You may have an AMD script outside your views you wish to make available as a module.
 
-The best way to accomplish this is by configuring the path to the file using the
-``views.php`` file in the root of your plugin:
+The best way to accomplish this is by configuring the path to the file using the views section of the
```
[the remark this answers is not in the document]
Comment (jdalsem) [reply docs/guides/javascript.rst]: fixed

File: docs/guides/javascript.rst [hunk 3]
```diff
@@ -168,15 +168,17 @@ Setting the URL of a module
...
 
 .. code-block:: php
 
-    <?php // views.php
+    <?php // elgg-plugin.php
     return [
-        'default' => [
-            'underscore.js' => 'vendor/bower-asset/underscore/underscore.min.js',
+        'views' => [
+	        'default' => [
```
[the remark this answers is not in the document]
Comment (jdalsem) [reply docs/guides/javascript.rst]: think so

File: docs/guides/javascript.rst [hunk 4]
```diff
@@ -185,10 +187,12 @@ you can use something like this instead:
 
 .. code-block:: php
 
-    <?php // views.php
+    <?php // elgg-plugin.php
     return [
-        'default' => [
-            'underscore.js' => __DIR__ . '/bower_components/underscore/underscore.min.js',
+        'views' => [
+	        'default' => [
```
[the remark this answers is not in the document]
Comment (jdalsem) [reply docs/guides/javascript.rst]: think so

File: docs/guides/plugins.rst [hunk 5]
```diff
@@ -123,12 +125,51 @@ Besides magic constants like ``__DIR__``, its return value should not change. Th
...
+				],
+			],
+			'likes:is_likable' => [
+				'object:blog' => [
+					'Elgg\Values::getTrue' => [],
+				],
+			],
+			'usersettings:save' => [
+				'user' => [
+					'_elgg_save_notification_user_settings' => ['unregister' => true],
```
[the remark this answers is not in the document]
Comment (jdalsem) [reply docs/guides/plugins.rst]: to show there are different ways of doing it

File: docs/guides/plugins.rst [hunk 6]
```diff
@@ -123,12 +125,51 @@ Besides magic constants like ``__DIR__``, its return value should not change. Th
...
+				],
+			],
+			'create' => [
+				'relationship' => [
+					'_elgg_send_friend_notification' => [],
+				],
+			],
+			'log' => [
+				'systemlog' => [
+					'system_log_default_logger' => ['unregister' => true],
```
[the remark this answers is not in the document]
Comment (jdalsem) [reply docs/guides/plugins.rst]: to show there are different ways of doing it

File: engine/tests/phpunit/unit/ElggPluginUnitTest.php (hunk 1 above)
Comment (jdalsem) [reply engine/tests/phpunit/unit/ElggPluginUnitTest.php]: they are not present in the test file... just added these as they were not tested... testing reading more sections makes no sense

**Pull request merged**